[SPMD] Remove ddp_gpu_size field from SyncBatchNorm

wayi1 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #56425 [SPMD] Remove _specify_ddp_gpu_num method
* **#55946 [SPMD] Remove ddp_gpu_size field from SyncBatchNorm**

As `ddp_gpu_size` field of `SyncBatchNorm` will always be 1 for GPU modules, remove this field and the relevant code.

Differential Revision: [D27746021](https://our.internmc.facebook.com/intern/diff/D27746021/)

Commit:
[SPMD] Remove _specify_ddp_gpu_num from SyncBatchNorm

As `ddp_gpu_size` field of `SyncBatchNorm` will always be 1 for GPU modules, remove this field and the relevant code.

Differential Revision: [D27746021](https://our.internmc.facebook.com/intern/diff/D27746021/)

[ghstack-poisoned]

zhaojuanmao commented on torch/nn/parallel/distributed.py:
```diff
@@ -1205,7 +1205,6 @@ def _passing_sync_batchnorm_handle(self, module_copies):
                     assert (
                         self.device_type != "cpu"
                     ), "SyncBatchNorm layers only work with GPU modules"
-                    layer._specify_ddp_gpu_num(1)
```
if I understand it correctly, _specify_ddp_gpu_num(1) is kind of passing handle to syncBatchNorm, letting syncBatchNorm know that the syncBatchNorm is working with DDP.

SciPioneer replied:
This really depends on whether we want to remove `ddp_gpu_size` field, which will always be 1 if not None.

Do you think we should still keep this field?

zhaojuanmao commented on torch/nn/modules/batchnorm.py:
```diff
@@ -684,9 +683,6 @@ def forward(self, input: Tensor) -> Tensor:
                 input, running_mean, running_var, self.weight, self.bias,
                 bn_training, exponential_average_factor, self.eps)
         else:
-            if not self.ddp_gpu_size:
```
looks like it is intentionally restricting SyncBatchNorm to work with DDP only. feel it is not hard to keep it?

otherwise, we can completely kill _specify_ddp_gpu_num() and self.ddp_gpu_size

SciPioneer replied:
I thought it was used for restricting DDP SPSD mode. Since now SPMD is gone, I am thinking of completely killing `_specify_ddp_gpu_num()` and `self.ddp_gpu_size`.

For now, `_specify_ddp_gpu_num()` is kept for BC only, and it only checks if the input module is a GPU module not CPU module.

zhaojuanmao replied:
I see, I think it would be good to provide another way to check input module is a GPU module or not before completely retiring _specify_ddp_gpu_num() and self.ddp_gpu_size in the followup diffs. 

Thanks!

SciPioneer replied:
The only 2 callers of `_specify_ddp_gpu_num`, both checked whether the module is a GPU module before invoking this function. Therefore, `_specify_ddp_gpu_num` becomes totally useless now. Will remove it in the next PR.

rohan-varma reviewed (commented):
I think I'm a bit unclear on the original `_specify_ddp_gpu_num`, is it restricting any caller to have to be running within distributed/ddp context, or is it explicitly guarding against SPMD mode (seems to be the latter)?

If it is indeed the latter, then this PR should be good to go, but maybe per @zhaojuanmao's comment we should somehow disable/throw on SyncBN being used outside of DDP? Not sure what currently happens in this case.

Commit:
Update on "[SPMD] Remove ddp_gpu_size field from SyncBatchNorm"


As `ddp_gpu_size` field of `SyncBatchNorm` will always be 1 for GPU modules, remove this field and the relevant code.

Differential Revision: [D27746021](https://our.internmc.facebook.com/intern/diff/D27746021/)

[ghstack-poisoned]
